iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResource.java

```diff
@@ -0,0 +1,217 @@
+
+  public File getFile() {
+    return hardlinkOrCopiedFile;
+  }
+
+  ///////////////////// Reference Count /////////////////////
+
+  public int getReferenceCount() {
+    return referenceCount.get();
+  }
+
+  public int increaseAndGetReference() {
+    return referenceCount.addAndGet(1);
+  }
+
+  public int decreaseAndGetReference() {
+    final int finalReferenceCount = referenceCount.addAndGet(-1);
+    if (finalReferenceCount == 0) {
+      lastUnpinToZeroTime.set(System.currentTimeMillis());
+    }
+    if (finalReferenceCount < 0) {
+      LOGGER.warn("PipeTsFileResource's reference count is decreased to below 0.");
+    }
+    return finalReferenceCount;
+  }
+
+  public synchronized boolean closeIfOutOfTimeToLive() throws IOException {
+    if (referenceCount.get() <= 0
+        && (deviceMeasurementsMap == null // Not cached yet.
+            || System.currentTimeMillis() - lastUnpinToZeroTime.get()
+                > TSFILE_MIN_TIME_TO_LIVE_IN_MS)) {
+      close();
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public synchronized void close() throws IOException {
+    if (deviceMeasurementsMap != null) {
+      deviceMeasurementsMap = null;
+    }
+
+    if (deviceIsAlignedMap != null) {
+      deviceIsAlignedMap = null;
+    }
+
+    if (measurementDataTypeMap != null) {
+      measurementDataTypeMap = null;
+    }
+
+    if (allocatedMemoryBlock != null) {
+      allocatedMemoryBlock.close();
+      allocatedMemoryBlock = null;
+    }
+
+    Files.deleteIfExists(hardlinkOrCopiedFile.toPath());
+
+    LOGGER.info("PipeTsFileResource: Closed tsfile {} and cleaned up.", hardlinkOrCopiedFile);
+  }
+
+  //////////////////////////// Cache Getter ////////////////////////////
+
+  public synchronized Map<String, List<String>> tryGetDeviceMeasurementsMap() throws IOException {
+    if (deviceMeasurementsMap == null && isTsFile) {
+      tryCacheObjects();
+    }
+    return deviceMeasurementsMap;
+  }
+
+  public synchronized Map<String, Boolean> tryGetDeviceIsAlignedMap() throws IOException {
+    if (deviceIsAlignedMap == null && isTsFile) {
+      tryCacheObjects();
+    }
+    return deviceIsAlignedMap;
+  }
+
+  public synchronized Map<String, TSDataType> tryGetMeasurementDataTypeMap() throws IOException {
+    if (measurementDataTypeMap == null && isTsFile) {
+      tryCacheObjects();
+    }
+    return measurementDataTypeMap;
+  }
+
+  synchronized boolean tryCacheObjects() throws IOException {
+    if (!isTsFile) {
+      return false;
+    }
+
+    if (allocatedMemoryBlock != null) {
+      // This means objects are already cached.
+      return true;
+    }
+
+    // See if pipe memory is sufficient to be allocated for TsFileSequenceReader.
+    // Only allocate when pipe memory used is less than 50%, because memory here
+    // is hard to shrink and may consume too much memory.
+    allocatedMemoryBlock =
+        PipeResourceManager.memory()
+            .forceAllocateIfSufficient(
```

Comment:
Do we still need to allocate a memory for the reader in the circumstance that it's temporary and can be GCed?

Reply:
Maybe we don't need to actually allocate, but it's still useful to check if used pipe memory is below threshold.